libs/core/algorithms/include/hpx/parallel/algorithms/nth_element.hpp

```diff
@@ -0,0 +1,411 @@
+            else
+                nth_element_seq(c_last + 1, nth, end, level - 1, comp, proj);
+
+            return;
+        };
+
+        template <typename Iter>
+        struct nth_element : public detail::algorithm<nth_element<Iter>, Iter>
+        {
+            nth_element()
+              : nth_element::algorithm("nth_element")
+            {
+            }
+
+            template <typename ExPolicy, typename RandomIt, typename Sent,
+                typename Pred, typename Proj>
+            static RandomIt sequential(ExPolicy, RandomIt first, RandomIt nth,
+                Sent last, Pred&& pred, Proj&& proj)
+            {
+                auto end = detail::advance_to_sentinel(first, last);
+
+                auto nelem = end - first;
+                if (nelem == 0)
+                    return first;
+
+                uint32_t level = detail::nbits64(nelem) * 2;
+                detail::nth_element_seq(first, nth, end, level,
+                    std::forward<Pred>(pred), std::forward<Proj>(proj));
+
+                return end;
+            }
+
+            template <typename ExPolicy, typename RandomIt, typename Sent,
+                typename Pred, typename Proj>
+            static typename util::detail::algorithm_result<ExPolicy,
+                RandomIt>::type
+            parallel(ExPolicy&& policy, RandomIt first, RandomIt nth, Sent last,
+                Pred&& pred, Proj&& proj)
+            {
+                using value_type =
+                    typename std::iterator_traits<RandomIt>::value_type;
+
+                RandomIt partitionIter, return_last;
+
+                if (first == last)
+                {
+                    return util::detail::algorithm_result<ExPolicy,
+                        RandomIt>::get(std::move(first));
+                }
+
+                if (nth == last)
+                {
+                    return util::detail::algorithm_result<ExPolicy,
+                        RandomIt>::get(std::move(nth));
+                }
+
+                try
+                {
+                    RandomIt last_iter =
+                        detail::advance_to_sentinel(first, last);
+                    return_last = last_iter;
+
+                    while (first != last_iter)
+                    {
+                        auto n = detail::distance(first, last_iter);
+
+                        // get random pivot index
+                        auto pivotIndex = std::rand() % n;
+                        // swap first and pivot element
+#if defined(HPX_HAVE_CXX20_STD_RANGES_ITER_SWAP)
+                        std::ranges::iter_swap(first, first + pivotIndex);
+#else
+                        std::iter_swap(first, first + pivotIndex);
+#endif
+
+                        auto partitionResult =
+                            hpx::parallel::v1::detail::partition<RandomIt>()
+                                .call(
+                                    std::forward<ExPolicy>(policy), first + 1,
```

Comment:
Wouldn't it be better to always use a synchronous policy here (since we wait for it below anyways)?

Reply:
Yeah, so do I always pass in an hpx::execution::par(but then what if the user passes in a par_unseq)? Is there a way to "convert" a parallel asynchronous policy to it's corresponding synchronous one?

Reply:
We don't have anything ready for this, I believe. One way would be to do:
```
auto non_task_policy = par.on(policy.executor()).with(policy.parameters());
```
This will break with `par_unseq` and similar, however, So it might be better to create a trait that can be specialized for the task policies and that returns the given policy unchanged for all non-task policies.

Reply:
Oh ok got it, thanks!